bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/graphics/Resource.java

```diff
@@ -142,6 +138,7 @@ void destroyHandlesExcept(Set<Integer> zoomLevels) {
  * This method does nothing if the resource is already disposed.
  */
 public void dispose() {
+	if (tracker != null) tracker.reporting.set(false);
```

Comment:
And manually run the clean operation here so that it is cleared out of the cleaner list immediately

```suggestion
	if (tracker != null) tracker.reporting.set(false);
	if (cleanable != null) cleanable.clean(); // maybe after isDisposed check and maybe assign it null too?
```

Reply:
Isn't the Runnable invoked directly in the UI thread when we do this? I mean it doesn't have that much overhead, but we can do this 100% lazy in the cleaner thread.

Reply:
The runnable is done in the UI thread in that case, but by doing it there the disposal is complete as the cleanable is removed from the cleaner and no new thread is needed at gc time either.